core/trino-main/src/main/java/io/trino/operator/ChangeOnlyUpdatedColumnsMergeProcessor.java

```diff
@@ -83,6 +91,20 @@ public Page transformPage(Page inputPage)
 
         Page result = new Page(builder.toArray(Block[]::new));
 
+        for (int nonNullColumnChannel : nonNullColumnChannels) {
+            Block block = result.getBlock(nonNullColumnChannel);
+            if (block.mayHaveNull()) {
+                for (int position = 0; position < positionCount; position++) {
+                    long operation = TINYINT.getLong(operationChannelBlock, position);
+                    if (operation == INSERT_OPERATION_NUMBER || operation == UPDATE_OPERATION_NUMBER) {
+                        if (block.isNull(position)) {
+                            throw new TrinoException(NULL_TREATMENT_NOT_ALLOWED, "Assigning NULL to non-null MERGE target table column " + nonNullColumnChannel);
```

Comment:
Use `CONSTRAINT_VIOLATION` and also change `io.trino.sql.planner.LogicalPlanner#createNullNotAllowedFailExpression` to use that in a separate commit.

Reply:
Don't include the column channel number in the error message. We'd want to provide the column name if possible, but that's likely too complicated to get the name here.

Reply:
@electrum see @homar 's https://github.com/trinodb/trino/pull/13462 on how to do validation and report offending column name
@homar is working on UPDATE null checks now

Reply:
Changed, and additional commit added.